engine.go

```diff
@@ -115,6 +124,46 @@ func (w workflowEngine) Parse(
 	}, nil
 }
 
+// StepWorkflowPaths finds all the file paths to workflows referenced
+// in a workflow's steps. The key for each found workflow file is the
+// file path as it is written in this workflow.
+func StepWorkflowPaths(wf *workflow.Workflow) map[string]string {
+	stepFilePaths := map[string]string{}
+	for _, stepData := range wf.Steps {
+		stepDataMap, ok1 := stepData.(map[any]any)
+		if ok1 {
+			kind, ok1 := stepDataMap["kind"]
+			if ok1 {
+				kindString := kind.(string)
+				if kindString == "foreach" {
+					subworkflowPath := stepDataMap["workflow"]
+					subworkflowPathString := subworkflowPath.(string)
+					stepFilePaths[subworkflowPathString] = subworkflowPathString
+				}
+			}
+		}
+	}
+	return stepFilePaths
+}
+
+// SubworkflowCache creates a file cache of the sub-workflows referenced
+// in this workflow using rootDir as a context.
+func SubworkflowCache(wf *workflow.Workflow, rootDir string) (loadfile.FileCache, error) {
+	stepWorkflowPaths := StepWorkflowPaths(wf)
+	if len(stepWorkflowPaths) == 0 {
+		return nil, nil
+	}
+	stepWorkflowFileCache, err := loadfile.NewFileCacheUsingContext(rootDir, stepWorkflowPaths)
```

Comment:
OK, so this suggests that the handling of `rootDir` when you merge is irrelevant as they'll always be the same. I might consider whether to enforce that by verifying identity during the merge. *If* you managed to merge a different root dir, that means all the file path resolutions for subsequent sub-workflows would use the changed root, which could break things in confusing ways if you change the root silently.

Reply:
Yeah, this is more of an issue if we ever allow the user to submit more than one context directory for `rootDir`.

Reply:
It's probably better solved with your previous idea about nested file caches.

Reply:
Sure: right now it's not a problem. But since we don't expect it to differ, and code may *break* if it does, it'd be much better to catch and diagnose that now rather than wait to track down the cause of a breakage later!

Reply:
Won't do. I think y'all are trying to solve a problem that doesn't exist in the current domain. I think this issue is better addressed when we can expect to deal with multiple context or root directories.

Reply:
What we're recommending is an attempt to future-proof the code.

At present, the code expects and possibly depends on all workflows having the same root directory.  However, the code as it stands overwrites the root directory set by the main workflow when the subworkflows are merged in, and it does so without any check.  As you say, in the current domain, there is no problem because we're pretty darned sure that the value is overwritten by the same value.  However, given that surety, why bother doing the write at all since it has no effect?

I concur that we don't need to do much now to address a problem that we won't have until the future.  However, an ounce of prevention being worth a pound of cure, it is easy to add a check now (while we are aware of the issue) which will ensure that the problem is properly addressed when the time comes.  I expect that our future-selves won't mind, and they might even thank us.

Reply:
I disagree, but as we both gave approval on the basis that it "ought to work as-is for now", I won't belabor the point too much more. Actively handling a case in a way that you know is going to be wrong if it ever matters (by quietly overwriting the root) doesn't seem very robust. If you want the code to assume they're always the same there's no need to worry about the new root in each merged cache: just retain the original value. (Though at least a log message if they ever differ might be appreciated some day.)

Reply:
[It cracks me up that GitHub displays the comments in the order that they were created (i.e., timestamped) and not the order in which they were posted. That is, Dave's post from 28 minutes before this one is disagreeing with Matt's post from an hour ago, not with my post from 33 minutes ago, which appears before Dave's although it was actually posted (committed) after his. 🙂]

Reply:
I don't think the extra work is only an ounce. It seems like the current implementation would need a significant change.

Reply:
I think that all we are suggesting is adding a couple of lines [here](https://github.com/arcalot/arcaflow-engine/blob/08505a8dd6c2a480064c7a36f3267c69638e58bf/loadfile/loadfile.go#L164):
```go
		}
		if rootDir != "" and rootDir != fc.RootDir() {
			panic("Subworkflow has mismatched root directory.")
		}
		rootDir = fc.RootDir()
	}
```
Or, I might prefer this, instead:
```go
		}
		if rootDir == "" {
			rootDir = fc.RootDir()
		} else if rootDir != fc.RootDir()
			panic("Subworkflow has mismatched root directory.")
		}
	}
```

Reply:
I suppose I'm amenable, but it would have to use `fmt.Errorf()` and change `MergeFileCaches()` to return `(FileCache, error)`. I'll submit a change shortly.

Reply:
(This is @dbutenhof's conversation to resolve....)